### PR TITLE
Cleanup of enchantments

### PR DESCRIPTION
**Pull request by parlough**
[SpongeAPI](https://github.com/SpongePowered/SpongeAPI/pull/1687) | **SpongeCommon** | [SpongeVanilla](https://github.com/SpongePowered/SpongeVanilla/pull/361) | [SpongeForge](https://github.com/SpongePowered/SpongeForge/pull/1847)

Read API PR for more information.

### CODE REVIEW

File: src/main/java/org/spongepowered/common/mixin/core/item/MixinEnchantment.java [hunk 1]
```diff
@@ -108,13 +109,12 @@ public boolean canBeAppliedByTable(ItemStack stack) {
...
     @Override
-    public boolean isCompatibleWith(Enchantment ench) {
+    public boolean isCompatibleWith(EnchantmentType ench) {
         return canApplyTogether((net.minecraft.enchantment.Enchantment) ench);
     }
 
-    @Intrinsic
     public String enchantment$getName() {
-        return shadow$getName();
+        return I18n.translateToLocal(shadow$getName());
```
Comment (liach) [inline src/main/java/org/spongepowered/common/mixin/core/item/MixinEnchantment.java]: Is this change intended?

File: src/main/java/org/spongepowered/common/registry/type/data/KeyRegistryModule.java [hunk 2]
```diff
@@ -436,7 +436,7 @@ public void registerDefaults() {
 
         this.fieldMap.put("last_date_played", makeSingleKey(TypeTokens.INSTANT_TOKEN, TypeTokens.INSTANT_VALUE_TOKEN, of("LastTimePlayed"), "sponge:last_time_played", "Last Time Played"));
 
-        this.fieldMap.put("hide_enchantments", makeSingleKey(TypeTokens.BOOLEAN_TOKEN, TypeTokens.BOOLEAN_VALUE_TOKEN, of("HideEnchantments"), "sponge:hide_enchantments", "Hide Enchantments"));
+        this.fieldMap.put("hide_enchantments", makeSingleKey(TypeTokens.BOOLEAN_TOKEN, TypeTokens.BOOLEAN_VALUE_TOKEN, of("HideEnchantments"), "sponge:hide_enchantments", "Hide EnchantmentTypes"));
```
Comment (liach) [inline src/main/java/org/spongepowered/common/registry/type/data/KeyRegistryModule.java]: This one is not that much...

File: src/main/java/org/spongepowered/common/registry/type/data/KeyRegistryModule.java [hunk 3]
```diff
@@ -412,7 +412,7 @@ public void registerDefaults() {
 
         this.fieldMap.put("persists", makeSingleKey(TypeTokens.BOOLEAN_TOKEN, TypeTokens.BOOLEAN_VALUE_TOKEN, of("Persists"), "sponge:persists", "Persists"));
 
-        this.fieldMap.put("stored_enchantments", makeListKey(TypeTokens.LIST_ITEM_ENCHANTMENT_TOKEN, TypeTokens.LIST_ITEM_ENCHANTMENT_VALUE_TOKEN, of("StoredEnchantments"), "sponge:stored_enchantments", "Stored Enchantments"));
+        this.fieldMap.put("stored_enchantments", makeListKey(TypeTokens.LIST_ITEM_ENCHANTMENT_TOKEN, TypeTokens.LIST_ITEM_ENCHANTMENT_VALUE_TOKEN, of("StoredEnchantments"), "sponge:stored_enchantments", "Stored EnchantmentTypes"));
```
Comment (liach) [inline src/main/java/org/spongepowered/common/registry/type/data/KeyRegistryModule.java]: I think you should change these names.

File: src/main/java/org/spongepowered/common/registry/type/event/DamageModifierTypeRegistryModule.java [hunk 4]
```diff
@@ -57,15 +57,15 @@
...
+        this.modifierTypeMap.put("weapon_enchantment", new SpongeDamageModifierType("Weapon EnchantmentType", "weapon_enchantment"));
         this.modifierTypeMap.put("offensive_potion_effect", new SpongeDamageModifierType("Offensive PotionEffect", "offensive_potion_effect"));
         this.modifierTypeMap.put("defensive_potion_effect", new SpongeDamageModifierType("Defensive PotionEffect", "defensive_potion_effect"));
         this.modifierTypeMap.put("negative_potion_effect", new SpongeDamageModifierType("Negative PotionEffect", "negative_potion_effect"));
         this.modifierTypeMap.put("hard_hat", new SpongeDamageModifierType("Hard Hat", "hard_hat"));
         this.modifierTypeMap.put("shield", new SpongeDamageModifierType("Shield", "shield"));
         this.modifierTypeMap.put("blocking", this.modifierTypeMap.get("shield"));
         this.modifierTypeMap.put("armor", new SpongeDamageModifierType("Armor", "armor"));
-        this.modifierTypeMap.put("armor_enchantment", new SpongeDamageModifierType("Armor Enchantment", "armor_enchantment"));
+        this.modifierTypeMap.put("armor_enchantment", new SpongeDamageModifierType("Armor EnchantmentType", "armor_enchantment"));
```
Comment (liach) [inline src/main/java/org/spongepowered/common/registry/type/event/DamageModifierTypeRegistryModule.java]: Not these two

File: testplugins/src/main/java/org/spongepowered/test/CustomInventoryTest.java [hunk 5]
```diff
@@ -67,6 +69,20 @@
 public class CustomInventoryTest {
 
     @Listener
+    public void onBlockInteract(InteractBlockEvent event, @Root Player player) {
```
Comment (parlough) [inline testplugins/src/main/java/org/spongepowered/test/CustomInventoryTest.java]: This was for some testing, will remove this later.

File: src/main/java/org/spongepowered/common/mixin/core/item/MixinEnchantment.java (hunk 1 above)
Comment (parlough) [reply src/main/java/org/spongepowered/common/mixin/core/item/MixinEnchantment.java]: Yeah as the getName method previously returned the translation key of "enchantment." + name. For example "enchantment.waterWalker" or similar.

**Comment by parlough** [comment]
One second, will fix the issue...

File: src/main/java/org/spongepowered/common/data/util/DataQueries.java [hunk 6]
```diff
@@ -64,8 +64,8 @@
     public static final DataQuery VELOCITY_Y = of("Y");
     public static final DataQuery VELOCITY_Z = of("Z");
 
-    // Enchantments
-    public static final DataQuery ENCHANTMENT_ID = of("Enchantment");
```
Comment (gabizou) [inline src/main/java/org/spongepowered/common/data/util/DataQueries.java]: This may need to be deprecated legacy versioned

File: src/main/java/org/spongepowered/common/item/enchantment/SpongeEnchantmentBuilder.java [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+    @Nullable private EnchantmentType enchantmentType;
+    @Nullable private Integer level;
+
+    public SpongeEnchantmentBuilder() {
+        super(Enchantment.class, 1);
+    }
+
+    @Override
+    public Enchantment.Builder from(final Enchantment value) {
+        this.enchantmentType = value.getType();
```
Comment (gabizou) [inline src/main/java/org/spongepowered/common/item/enchantment/SpongeEnchantmentBuilder.java]: Needs a null check.

File: src/main/java/org/spongepowered/common/item/enchantment/SpongeEnchantmentBuilder.java [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+    public Enchantment build() {
+        checkState(this.enchantmentType != null, "The enchantment type must be set!");
+        checkState(this.level != null, "The level of the enchantment must be set!");
+        return new SpongeEnchantment(this.enchantmentType, this.level);
+    }
+
+    @Override
+    protected Optional<Enchantment> buildContent(DataView container) throws InvalidDataException {
+        checkNotNull(container, "The data view cannot be null!");
+        if (container.contains(Queries.ENCHANTMENT_ID, Queries.LEVEL)) {
```
Comment (gabizou) [inline src/main/java/org/spongepowered/common/item/enchantment/SpongeEnchantmentBuilder.java]: Invert the if block, style change, I know, just preference. Fail fast.

File: src/main/java/org/spongepowered/common/item/enchantment/SpongeEnchantmentBuilder.java [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+            final String id = DataUtil.getData(container, Queries.ENCHANTMENT_ID, String.class);
+            final int level = DataUtil.getData(container, Queries.LEVEL, Integer.class);
+            final Optional<EnchantmentType> enchantmentType = SpongeImpl.getRegistry().getType(EnchantmentType.class, id);
```
Comment (gabizou) [inline src/main/java/org/spongepowered/common/item/enchantment/SpongeEnchantmentBuilder.java]: SpongeImpl vs Sponge?

File: src/main/java/org/spongepowered/common/mixin/core/item/MixinEnchantment.java [hunk 10]
```diff
@@ -44,8 +44,8 @@
 
 @NonnullByDefault
 @Mixin(net.minecraft.enchantment.Enchantment.class)
-@Implements(@Interface(iface = Enchantment.class, prefix = "enchantment$"))
-public abstract class MixinEnchantment implements Enchantment, IMixinEnchantment {
+@Implements(@Interface(iface = EnchantmentType.class, prefix = "enchantment$"))
+public abstract class MixinEnchantment implements IMixinEnchantment {
```
Comment (gabizou) [inline src/main/java/org/spongepowered/common/mixin/core/item/MixinEnchantment.java]: Why remove it? It'll still perfectly implement just fine normally.

File: src/main/resources/META-INF/common_at.cfg [hunk 11]
```diff
@@ -20,7 +20,7 @@ public net.minecraft.command.EntitySelector func_82381_h(Ljava/lang/String;)Ljav
 public net.minecraft.crash.CrashReport field_71511_b # cause
 public net.minecraft.crash.CrashReportCategory field_85078_a # crashReport
 
-public net.minecraft.enchantment.Enchantment field_180311_a # enchantmentsList
+public net.minecraft.enchantmentType.Enchantment field_180311_a # enchantmentsList
```
Comment (gabizou) [inline src/main/resources/META-INF/common_at.cfg]: typo?

**Review by gabizou — CHANGES_REQUESTED** [review]
Small ages.

File: src/main/java/org/spongepowered/common/mixin/core/item/MixinEnchantment.java (hunk 10 above)
Comment (parlough) [reply src/main/java/org/spongepowered/common/mixin/core/item/MixinEnchantment.java]: `IMixinEnchantment` extends `Enchantment` so I didn't think it was needed.

Comment (gabizou) [reply src/main/java/org/spongepowered/common/mixin/core/item/MixinEnchantment.java]: Thinking about it further, the `IMixin` interfaces shouldn't extend the API interfaces for the basic understanding that they are only from implementation, The only reason why there's `IMixinEntity` and a few others doing this is because of Java compiler requirements with potentially ambiguous methods.

**Review by gabizou — CHANGES_REQUESTED** [review]
